DESIGN.md

```diff
@@ -0,0 +1,155 @@
+
+Though subject to change, it is currently defined as:
+
+```rust
+#[derive(Clone, Debug)]
+struct MatchTreeLevel<'a> {
+    accept: Option<usize>,
+    branches: Vec<(ByteSet, LevelBranch<'a>)>,
+}
+```
+
+where the `accept` field has the same semantics the like-named field in `MatchTree`.
+
+The `branches` field is defined in terms of a type-alias `LevelBranch<'_>`,
+outlined in the subsection found below
+
+#### LevelBranch
+
+The type-alias `LevelBranch` is defined as follows:
+
+```rust
+type LevelBranch<'a> = HashSet<(usize, Rc<Next<'a>>)>;
+```
+
+To understand the significance of this type, we make the following observation:
+
+In order for a `MatchTree` to be well-formed, it and all of its descendants
+must have mutually disjoint branching-predicates. That is, for all well-formed `tree : MatchTree`, it must be the case that for all values of `b : u8`, the iterator generated by `tree.branches.filter(|(s, _)| s.contains(b)))` must yield zero or one elements only.
+
+However, the same does not hold for `MatchTreeLevels`, as they represent the superposition of all choice-points at a common depth. If that depth is non-zero, the `MatchTree` may still be well-formed even if two choice-points within the `MatchTreeLevel` have some non-null intersection among the cartesian product of their respective lists of branching-predicates.
+
+For this reason, a `MatchTreeLevel` cannot represent a single, unambiguous descent into a follow-set for any given byte of input, but rather, must form a disjunction over the smallest set of mutually disjoint subsets of all the branching predicates of its constituent choice-points, such as in the following algorithm (in pseud-ocode):
+
+```none
+UnifySteps(Steps: [(FormatIx, MatchTreeStep)]):
+    Instantiate a variable X : MatchTreeLevel, with X.branches := [] and X.accept = None
+    For each element (Ix, Step) in Steps:
+        If Step.accept
+        Then
+            If X.accept ≣ None || X.accept ≣ Some(Ix)
+            Then
+                X.accept := Some(Ix)
+            Else
+                Return None
+        Else
+            For each element (BS, Next) in Step.branches:
+                Create a new temporary buffer NewBranches := []
+                For each element (BS0, Branch) in X.branches:
+                    Let Common := BS ∩ BS0
+                    If Common ≠ ∅
+                    Then
+                        Replace (BS0, Branch) with (Common, Branch ⊕ [(Ix, Next)]), mutably in-place
+                        Push (BS0 ∖ Common, Branch) to the end of NewBranches
+                        Remove all elements of Common from BS, mutably in-place
+                    Else
+                        Do nothing
+                If BS ≠ ∅
+                Then
+                    Push (BS, [(Ix, Next)]) to the end of X.branches
+                Else
+                    Do nothing
+                Append all elements of NewBranches to the end of X.branches
+    Return Some(X)
```

Comment:
I'm not sure if this pseudo-code is going to be clearer than just pointing to the Rust code in the module itself but I guess it provides an alternative view!

Reply:
I feel that the intent of the code is easier to capture in pseudo-code, but I agree with you morally. It just might be easier to get around details of potentially opaque APIs like ByteSet. There is also no single method in the Rust code that this represents, since it is a left-fold of `merge_step` with the calls to `merge_accept` and `merge_branch` inlined. But indeed, showing the Rust code might be about as effective as writing this out.